File: app/src/main/java/com/example/sunkai/heritage/folkListviewAdapter.java
package com.example.sunkai.heritage;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.util.LruCache;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import com.example.sunkai.heritage.ConnectWebService.HandleFolk;
import com.example.sunkai.heritage.Data.HandlePic;
import com.example.sunkai.heritage.Data.MySqliteHandler;
import com.example.sunkai.heritage.Data.folkData;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.List;

/**
 * Created by sunkai on 2017/3/3.
 */

public class folkListviewAdapter extends BaseAdapter {
    private Context context;
    private List<folkData> datas;
    folkFragment folkFragment;
    private ListView listView;
    private LruCache<Integer,Bitmap> lruCache;
    public folkListviewAdapter(Context context,folkFragment folkFragment){
        this.context=context;
        int maxSize=(int)Runtime.getRuntime().maxMemory();
        int avilibleMemory=maxSize/16;
        lruCache=new LruCache<Integer,Bitmap>(avilibleMemory){
            protected int sizeOf(Integer key, Bitmap bitmap) {
                return bitmap.getByteCount();
            }
        };
        this.folkFragment=folkFragment;
        if(this.folkFragment!=null) {
            setProgress(true);
        }
        //获取folk的文本信息
        new getInformation().execute();
    }
    public int getCount(){
        return datas==null?0:datas.size();
    }
    public Object getItem(int position) {
        return datas.get(position);
    }
    public long getItemId(int position) {
        return position;
    }
    public View getView(int position, View convertView, ViewGroup parent){
        if(listView==null)
            listView=(ListView)parent;
        Holder vh;
        if(convertView==null) {
            LayoutInflater inflater = LayoutInflater.from(context);
            convertView = inflater.inflate(R.layout.folk_listview_layout, null);
            vh = new Holder();
            vh.v1 = (TextView) convertView.findViewById(R.id.list_text);
            vh.v2 = (TextView) convertView.findViewById(R.id.list_location);
            vh.v3 = (TextView) convertView.findViewById(R.id.list_title);
            vh.v4 = (ImageView) convertView.findViewById(R.id.list_img);
            vh.v5=(TextView)convertView.findViewById(R.id.list_divide);
            convertView.setTag(vh);
        }
        else
        {
            vh=(Holder)convertView.getTag();
        }
        folkData data=datas.get(position);
        String content=data.getContent();
        String location=data.getLocation();
        String title=data.getTitle();
        String divide=data.getDivide();
        vh.v1.setText("        " + content);
        vh.v2.setText(location);
        vh.v3.setText(title);
        vh.v4.setImageResource(R.drawable.empty_background);
        vh.v4.setTag(data.getId());
        Bitmap bitmap=lruCache.get(data.getId());
        if(bitmap==null){
            new getFolkImage(data.getId()).execute();
        }else{
            vh.v4.setImageBitmap(bitmap);
        }
        vh.v5.setText(divide);
        return convertView;
    }
    public List<folkData> getDatas(){
        return this.datas;
    }
    public void setProgress(boolean show){
        if(show){
            folkFragment.loadProgress.setVisibility(View.VISIBLE);
            folkFragment.setWidgetEnable(!show);
        }
        else{
            folkFragment.loadProgress.setVisibility(View.GONE);
            folkFragment.setWidgetEnable(!show);
        }
    }
    public void setNewDatas(List<folkData> datas){
        this.datas=datas;
        notifyDataSetChanged();
    }
    public List<folkData> getData(){
        return datas;
    }
    public class Holder {
        TextView v1,v2,v3,v5;
        ImageView v4;
    }

    private class getInformation extends AsyncTask<Void,Void,Void>{
        @Override
        protected Void doInBackground(Void... voids) {
            datas=HandleFolk.GetFolkInforMation();
            folkFragment.isLoadData=true;
            folkFragment.setData(true,datas);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            notifyDataSetChanged();
            setProgress(false);
        }
    }

    private class getFolkImage extends AsyncTask<Void,Void,Bitmap>{
        int id;
        SQLiteDatabase db=MySqliteHandler.INSTANCE.GetReadableDatabase();
        public getFolkImage(int id){this.id=id;}
        @Override
        protected Bitmap doInBackground(Void... voids) {
            Bitmap bitmap=null;
            String table="folk_image";
            String selection="id=?";
            String[] selectionArgs=new String[]{String.valueOf(id)};
            Cursor cursor=db.query(table,null,selection,selectionArgs,null,null,null);
            cursor.moveToFirst();
            byte[] img=null;
            if(!cursor.isAfterLast()){
                int imageIndex=cursor.getColumnIndex("image");
                img=cursor.getBlob(imageIndex);
                cursor.close();
                if(img!=null) {
                    InputStream in=new ByteArrayInputStream(img);
                    bitmap=HandlePic.handlePic(context,in,0);
                    lruCache.put(id,bitmap);
                    return bitmap;
                }
            }
            img=HandleFolk.GetFolkImage(id);
            if(img==null)
                return null;
            InputStream in=new ByteArrayInputStream(img);
            ContentValues contentValues=new ContentValues();
            contentValues.put("id",id);
            contentValues.put("image",img);
            db= MySqliteHandler.INSTANCE.GetWritableDatabase();
            db.insert(table,null,contentValues);
            bitmap=HandlePic.handlePic(context,in,0);
            lruCache.put(id,bitmap);
            return bitmap;
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            if(bitmap!=null){
                ImageView imageView=(ImageView)listView.findViewWithTag(id);
                if(imageView!=null){
                    imageView.setImageBitmap(bitmap);
                }
            }
        }
    }



}

File: app/src/main/java/com/example/sunkai/heritage/Adapter/FolkListviewAdapter.kt
package com.example.sunkai.heritage.Adapter

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.graphics.Bitmap
import android.os.AsyncTask
import android.util.LruCache
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import com.bumptech.glide.Glide

import com.example.sunkai.heritage.ConnectWebService.HandleFolk
import com.example.sunkai.heritage.Data.HandlePic
import com.example.sunkai.heritage.Data.MySqliteHandler
import com.example.sunkai.heritage.Data.FolkData
import com.example.sunkai.heritage.Data.FolkDataLite
import com.example.sunkai.heritage.R
import com.example.sunkai.heritage.Fragment.FolkFragment
import com.example.sunkai.heritage.value.HOST

import java.io.ByteArrayInputStream

/**
 * Created by sunkai on 2017/3/3.
 *
 */

class FolkListviewAdapter(private val context: Context, internal var folkFragment: FolkFragment?) : BaseAdapter() {
    var datas: List<FolkDataLite>?=null
    private var listView: ListView? = null
    private val lruCache: LruCache<Int, Bitmap>

    init {
        val maxSize = Runtime.getRuntime().maxMemory().toInt()
        val avilibleMemory = maxSize / 16
        lruCache = object : LruCache<Int, Bitmap>(avilibleMemory) {
            override fun sizeOf(key: Int?, bitmap: Bitmap): Int {
                return bitmap.byteCount
            }
        }
        if (this.folkFragment != null) {
            setProgress(true)
        }
        //获取folk的文本信息
        getInformation().execute()
    }

    override fun getCount(): Int {
        return if(datas==null)
            0
        else
            datas!!.size
    }

    override fun getItem(position: Int): Any {
        return datas!![position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    @SuppressLint("InflateParams")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view:View
        if (listView == null)
            listView = parent as ListView
        val vh: Holder
        if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            view = inflater.inflate(R.layout.folk_listview_layout, null)
            vh = Holder()
            vh.v1 = view.findViewById<View>(R.id.list_text) as TextView
            vh.v2 = view.findViewById<View>(R.id.list_location) as TextView
            vh.v3 = view.findViewById<View>(R.id.list_title) as TextView
            vh.v4 = view.findViewById<View>(R.id.list_img) as ImageView
            vh.v5 = view.findViewById<View>(R.id.list_divide) as TextView
            view.tag = vh
        } else {
            view=convertView
            vh = convertView.tag as Holder
        }
        val data = datas!![position]
        val title = data.title
        val divide = data.divide
        vh.v3.text = title
        vh.v4.setImageResource(R.drawable.empty_background)
        Glide.with(context).load(HOST+data.img).into(vh.v4)
        vh.v5.text = divide
        return view
    }

    fun setProgress(show: Boolean) {
        if (show) {
            folkFragment!!.loadProgress.visibility = View.VISIBLE
            folkFragment!!.setWidgetEnable(!show)
        } else {
            folkFragment!!.loadProgress.visibility = View.GONE
            folkFragment!!.setWidgetEnable(!show)
        }
    }

    fun setNewDatas(datas: List<FolkDataLite>) {
        this.datas = datas
        notifyDataSetChanged()
    }

    fun getData(): List<FolkDataLite>? {
        return datas
    }

    inner class Holder {
        internal lateinit var v1: TextView
        internal lateinit var v2: TextView
        internal lateinit var v3: TextView
        internal lateinit var v5: TextView
        internal lateinit var v4: ImageView
    }

    inner class getInformation : AsyncTask<Void, Void, Boolean>() {
        override fun doInBackground(vararg voids: Void): Boolean {
            datas = HandleFolk.GetFolkInforMation()
            datas?.let {
                FolkFragment.isLoadData = true
                folkFragment?.setData(true, datas!!)
            }
            return true
        }

        override fun onPostExecute(result: Boolean) {
            if(result) {
                notifyDataSetChanged()
            }
            setProgress(false)
        }
    }
}

